src/main/java/org/jabref/logic/util/BracketedExpressionExpander.java

```diff
@@ -0,0 +1,1356 @@
+package org.jabref.logic.util;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Scanner;
+import java.util.StringJoiner;
+import java.util.StringTokenizer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jabref.logic.formatter.Formatters;
+import org.jabref.logic.formatter.casechanger.Word;
+import org.jabref.logic.layout.format.RemoveLatexCommandsFormatter;
+import org.jabref.model.cleanup.Formatter;
+import org.jabref.model.database.BibDatabase;
+import org.jabref.model.entry.AuthorList;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.FieldName;
+import org.jabref.model.entry.Keyword;
+import org.jabref.model.entry.KeywordList;
+import org.jabref.model.strings.StringUtil;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * @author saulius
+ * The BracketedExpressionExpander provides methods to expand bracketed expressions,
+ * such as [year]_[author]_[firstpage], using information from a provided BibEntry.
+ * The above-mentioned expression would yield 2017_Kitsune_123 when expanded using the
+ * BibTeX entry "@Article{ authors = {O. Kitsune}, year = {2017}, pages={123-6}}".
+ */
+public class BracketedExpressionExpander {
+
+    private static final Log LOGGER = LogFactory.getLog(BracketedExpressionExpander.class);
+
+    private static final String STARTING_CAPITAL_PATTERN = "[^A-Z]";
+    private static final int CHARS_OF_FIRST = 5;
+    private static final Pattern REGEX_PATTERN = Pattern.compile(".*\\(\\{([A-Z]+)\\}\\).*");
+
+    private final BibEntry bibentry;
+
+    public BracketedExpressionExpander() {
+        this.bibentry = null;
+    }
+
+    /**
+     * @param bibentry
+     */
+    public BracketedExpressionExpander(BibEntry bibentry) {
```

Comment:
I'm not a big fan of the semantics here. Instead I would propose that `BracketedExpressionExpander` actually accepts the pattern as a constructor argument and then provides a method `expand` that accepts a BibEntry as an argument and uses the contained information to expand the pattern. My proposal is a bit similar to the regular expression pattern class.
Moreover, I would then rename the `BracketedExpressionExpander` class to something like `GenericPattern` or `BacketedPattern` (actually I would prefer just `Pattern` but this name is already taken by the java regex pattern class).

What do you think?

Reply:
Thaks for the renaming suggestion! I agree, 'BracketedExpressionExpander' is a long and clumsy name. I would opt for 'BacketedPattern': 'GenericPattern', I think, is too generic ;), and Pattern is taken, as you say. I've refactored the class and tests to use 'BacketedPattern' and 'BacketedPattern.expand' instead of 'BacketedPattern.exoandBrackets'.

Reply:
As for semantics, I thing one can do both ways. I assumed that the the common situation will be to have a pattern that we want to expand, and then have an (abstract) expander that contains all necessary information (bibentry, preferenceses, databases, etc.). I also assumed that one may want to use the same expander for several patterns in a row; this last assumption does not hold for JabRef.

So, I could change the interface to public BracketedPattern(String pattern, BibEntry bibentry), or public BracketedPattern(String pattern).

I must admit however that to my taste it seems a terrible abuse of OO to make an object as a fancy way of passing another parameter to a function; i.e. having 'X x = new X(y); x.f(z)' instead of simply 'f(y,z)'. Functions should stay functions, and object should be "clever" values used in those functions, IMHO :).

In the above case, 'expandBrackets' seems to be more like a function that takes two arguments, a pattern (which is always a string) and the context information (which may be BibEntry, Database, DatabaseContext, depending on your needs, and may change in the future). Having BracketedPattern.expand( String pattern ) would allow flexibility in the future: one could use descendent classes of BracketedPattern (with different context information) everywhere where the '.expand( bracketedPattern )' method is called, without changing and interface. Or one might have an AbstractBracketedPattern interface/abstract class, and implement '.expand( bracketedPattern )' in different ways. 

Given the above considerations, the current interface, 'public BracketedPattern(BibEntry bibentry)' and 'String BracketedPattern.expand(String pattern)' seem the most flexible and future-proof.

What are your thoughts?

Reply:
You are right,
>'expandBrackets' seems to be more like a function that takes two arguments, a pattern (which is always a string) and the context information (which may be BibEntry, Database, DatabaseContext, depending on your needs, and may change in the future)

So in principle there are two logical ways to split these arguments to the `expandBrackets` function:
a) The pattern as an argument to the constructor and the context information as arguments to the `expandBrackets` method.
b) The other way around, that is the context information as arguments to the constructor and the pattern as an argument to the `expandBrackets` method.

You favor method b) while I would use a). There are good arguments for both ways. Lets make a deal: I'll present my reasoning for a) and if I can convince you, you change the code accordingly. If not then the code stays like it is right now and I give my +1 for merge 😸 

So:
- In principle, we can do a pre-processing of the pattern in the constructor and then reuse the `expand` method for different entries to save time.
- You can create the `BracketedPattern` from the string in some place (say in the preference class) and then pass it to other methods (for example, the rename cleanup operation). Then only the preference class needs to know how to deserialize the pattern from a string, while the other methods in the chain only care that the are given a `BracketedPattern`. (Background: there was a time, when strings where used for everything in JabRef: File names, entry field names and values, preferences, ... literally everywhere. Since about 2 year we are now trying to encapsulate more and more of these strings in special classes, e. g. `AuthorList` or dedicated preference classes like `LayoutPreferences`. I would like to see the same for the pattern.)
- If we should decide to change or replace the pattern format, one would only need to change the places where the conversion from string to the pattern happens (which is considerably less overhead then all the places where the pattern is expanded). 
- If a class is named `BracketedPattern` I would assume that it encapsulates a pattern. 😈